README.md
# LoggerMulticastBackend

An Elixir `Logger` backend that uses multicast UDP to deliver log messages.
Designed for headless embedded applications, it allows watching the
log over the local network.

## Installation

If [available in Hex](https://hex.pm/docs/publish), the package can be installed as:

  1. Add `logger_multicast_backend` to your list of dependencies in `mix.exs`:

    ```elixir
      def deps do
        [{:logger_multicast_backend, "~> 0.2.1"}]
      end
    ```

  2. Add backend to Logger

    ```elixir
      Logger.add_backend LoggerMulticastBackend
    ```

## Defaults

In your logger config, simply do something like this:

```elixir
  config :logger,
    backends: [ :console, LoggerMulticastBackend ],
    level: :debug,
    format: "$time $metadata[$level] $message\n"
```

or, add the backend at runtime:

```elixir
  Logger.add_backend LoggerMulticastBackend
```

Now, you'll have logging messages sent out on the default target multicast
address, which is 224,0,0,224:9999.

## Configuration

A `Logger` backend that uses multicast UDP to deliver log messages. Originally
designed for embedded applications, it allows easily watching the log of a
headless device on the local network.

in your logger config, simply do something like this:

```elixir
  config :logger,
    backends: [ :console, LoggerMulticastBackend ]
    level: :debug,
    format: "$time $metadata[$level] $message\n"
  config :logger_multicast_backend,
    target: {{224,0,0,220}, 514},
    interface: {192,168,1,50},
    metadata: [:line, :file]
```

or, at runtime:

```elixir
  Logger.add_backend LoggerMulticastBackend
  Logger.configure_backend LoggerMulticastBackend, [target: {{224,0,0,220}, 514}, interface: {192,168,1,50}]
```

LoggerMulticastBackend is configured when specified, and supports the following options:

`:target` - a tuple of the target unicast or multicast address and port. Defaults to: {{224,0,0,224}, 9999}

`:interface` - If the host has more then one network interfaces, specify which one to use by IP address. Defaults to: {0,0,0,0}

`:level` - the level to be logged by this backend. _Note that messages are first filtered by the general `:level` configuration in `:logger`_. Defaults to: :debug

`:format` - the format message used to print logs. Defaults to: "$time $metadata[$level] $levelpad$message\n"

`:metadata` - the metadata to be printed by $metadata. Defaults to: \[\] (no metadata)

CHANGELOG.md
# Change Log

## v0.3.0

Major re-architecture (again)

* Now uses `GenStage` for `Queue` and `Sender` portions.
* Supports specifying network adapter to bind if system has more then one.


## v0.2.0-dev

Major re-architecture

- Now handles port coming/going gracefully
- Caches all data sent when port is not open, and re-sends it when it is
- Keeps up to the last 1024 lines of log for a port that is closed
- Throttles UDP traffic to 10ms packet pacing
- Tolerates IP address changes properly
- Removed blocking send (speeds up logger overall)

## v0.0.1

Initial checkin, lots of bugs

config/test.exs
use Mix.Config

config :logger, backends: [LoggerMulticastBackend]

config :logger_multicast_backend,
  format: "$metadata[$level] $message\n"

test/listener.ex
defmodule LoggerMulticastBackendTest.Listener do
  use GenServer

  def start(args \\ []) do
    case GenServer.start_link __MODULE__, args, name: TestListener do
      {:error, {:already_started, pid}} ->
        {:ok, pid}
      ret ->
        ret
    end
  end

  def init(_args) do
    open_socket()
    {:ok, []}
  end

  def handle_info({:udp, _socket, _ip, _port, data}, state) do
    {:noreply, state ++ [data]}
  end

  def handle_call(:get_line, _from, []) do
    {:reply, nil, []}
  end
  def handle_call(:get_line, _from, state) do
    [h | t] = state
    {:reply, h, t}
  end

  defp open_socket do
    :gen_udp.open(9999, [
      :binary,
      {:active, true},
      {:recbuf, 65536},
  	  {:reuseaddr, true},
      {:add_membership, {{224,0,0,224}, {0,0,0,0}}}
    ])
  end
end

test/logger_multicast_backend_test.exs
Code.require_file "listener.ex", __DIR__
defmodule LoggerMulticastBackendTest do

  use ExUnit.Case
  require Logger

  setup do
    LoggerMulticastBackendTest.Listener.start
    Logger.configure_backend LoggerMulticastBackend, [format: "$metadata[$level] $message\n"]
    :ok
  end

  test "test debug message" do
    Logger.debug "A debug message"
    :timer.sleep 20
    line = GenServer.call TestListener, :get_line
    assert line == "[debug] A debug message\n"
  end

  test "test info message" do
    Logger.info "A info message"
    :timer.sleep 20
    line = GenServer.call TestListener, :get_line
    assert line == "[info] A info message\n"
  end

  test "info event with error level" do
    Logger.configure_backend LoggerMulticastBackend, [level: :error]
    Logger.info "A info message"
    :timer.sleep 20
    line = GenServer.call TestListener, :get_line
    assert line == nil
  end

end



mix.exs
defmodule LoggerMulticastBackend.Mixfile do

  use Mix.Project

  @version "0.3.0"

  def project, do: [
    app: :logger_multicast_backend,
    version: @version,
    elixir: "~> 1.0",
    description: description(),
    package: package(),
    deps: deps(),
    docs: [
            source_ref: "v#{@version}", main: "LoggerMulticastBackend",
            source_url: "https://github.com/cellulose/logger_multicast_backend",
            extras: [ "README.md", "CHANGELOG.md"]
          ]
  ]

  defp description, do: "Elixir Logger backend using Multicast UDP"

  defp package, do: [
    contributors: ["Garth Hitchens", "Chris Dutton"],
    licenses: ["MIT"],
    links: %{"GitHub" => "https://github.com/cellulose/logger_multicast_backend"},
    files: ~w(lib config) ++ ~w(README.md CHANGELOG.md LICENSE mix.exs)
  ]

  def application, do: [ applications: [:gen_stage, :logger] ]

  defp deps, do: [
    {:ex_doc, "~> 0.7", only: :dev},
    {:gen_stage, "~> 0.4"}
  ]
end


lib/logger_multicast_backend/queue.ex
defmodule LoggerMulticastBackend.Queue do
  @moduledoc """
  A `GenStage` `producer` that implements a FIFO queue that dispatches to a `consumer`
  each line individual based on the demand requested.
  """
  alias Experimental.GenStage
  use GenStage

  @max_queue_len 1024 # maximum lines to save in the queue

  @doc """
  Initilizes module with empty queue
  """
  def init(_) do
    {:producer, {:queue.new, 0}}
  end

  ## Callbacks

  @doc """
  Add logger line to send queue
  """
  def handle_cast({:enqueue, line}, {queue, demand}) do
    queue = :queue.in(line, queue)
    queue = case :queue.len(queue) do
      q when  q >= @max_queue_len ->
        #Discard oldest line if queue has reached max
        {{:value, _line}, queue} = :queue.out(queue)
        queue
      _ -> queue
    end
    dispatch_lines(queue, demand, [])
  end

  @doc """
  Flush the queue - remove all lines currently in queue and reset demand
  """
  def handle_cast(:flush, {_queue, _demand}) do
    #REVIEW: Make sure old queue will get garbage collected or this is a memory leak
    {:noreply, [], {:queue.new, 0}}
  end

  @doc "GenStage callback to handle demand request from consumer"
  def handle_demand(incoming_demand, {queue, demand}) do
    dispatch_lines(queue, incoming_demand + demand, [])
  end

  # Dispatches lines to consumer from the queue
  defp dispatch_lines(queue, demand, lines) do
    with d when d > 0 <- demand,
         {{:value, line}, queue} <- :queue.out(queue) do
      dispatch_lines(queue, demand - 1, [line | lines])
    else
      _ ->
        {:noreply, Enum.reverse(lines), {queue, demand}}
    end
  end
end


lib/logger_multicast_backend/sender.ex
defmodule LoggerMulticastBackend.Sender do

  @moduledoc """
  `GenStage` `Consumer` that implements the sending-server for the LoggerMulticastBackend.
  Provides a managed timer-based sender for multicast logs as well as UDP socket management.
  """

  alias Experimental.GenStage
  use GenStage

  @socket_retry_time  1000      # milliseconds between port open attempts
  @packet_pacer_time  10        # milliseconds between multicasts

  @socket_opts [:binary, {:broadcast, true}, {:active, false}, {:reuseaddr, true} ]

  @doc "Init `Consumer` with target to send lines one and interface to bind socket to"
  def init([target: target, interface: interface]) do
    state = %{target: target, socket: nil, interface: interface}
    {:consumer, state}
  end

  @doc "Handles lines (events) received from `GenStage` `Producer`. Sends out on timer-based spacing"
  def handle_events(events, _from, state) do
    for e <- events do
      Process.send_after(self, {:send, e}, @packet_pacer_time)
    end
    {:noreply, [], state}
  end

  def handle_call({:configure, target, interface}, _from, state) do
    if state.socket, do: :gen_udp.close state.socket
    opts = @socket_opts ++ [{:ip, interface}] |> List.flatten
    case :gen_udp.open(0, opts) do
      {:ok, socket} ->
        {:reply, :ok, [], %{state | target: target, interface: interface, socket: socket}}
      _ ->
        {:reply, :ok, [], %{state | target: target, interface: interface}}
    end
  end

  @doc false
  # Handles request to send line with no socket defined - attempt to open it
  def handle_info({:send, line}, %{socket: nil} = state) do
    opts = @socket_opts ++ [{:ip, state.interface}] |> List.flatten
    case :gen_udp.open(0, opts) do
      {:ok, socket} ->
        state =  %{state | socket: socket}
        send_line(line, state)
        {:noreply, [], state}
      _ ->
        Process.send_after(self, {:send, line}, @socket_retry_time)
        {:noreply, [], state}
    end
  end

  @doc false
  # handle a request to send line with a valid socket - attempt to write to it
  def handle_info({:send, line}, state) do
    {:noreply, [], send_line(line, state)}
  end

  @doc false
  def terminate(msg, state) do
    IO.write "#{__MODULE__} terminating due to: #{inspect msg}\n"
    if state.socket, do: :gen_udp.close state.socket
    :ok
  end

  #Sends line out existing socket
  defp send_line(line, %{socket: socket, target: {addr, port}} = state) do
    case :gen_udp.send(socket, addr, port, line) do
      :ok ->
        state
      _ ->
        :gen_udp.close socket
        %{state | socket: nil}
    end
  end

end


lib/logger_multicast_backend.ex
defmodule LoggerMulticastBackend do

  @moduledoc """
  A `Logger` backend that uses multicast UDP to deliver log messages. Originally
  designed for embedded applications, it allows easily watching the log of a
  headless device on the local network.

  in your logger config, simply do something like this:

  ```elixir
    config :logger,
      backends: [ :console, LoggerMulticastBackend ]
      level: :debug,
      format: "$time $metadata[$level] $message\n"
    config :logger_multicast_backend,
      target: {{224,0,0,220}, 514},
      interface: {192,168,1,50},
      metadata: [:line, :file]
  ```

  or, at runtime:

  ```elixir
    Logger.add_backend LoggerMulticastBackend
    Logger.configure_backend LoggerMulticastBackend, [target: {{224,0,0,220}, 514}, interface: {192,168,1,50}]
  ```

  LoggerMulticastBackend is configured when specified, and supports the following options:

  :target - a tuple of the target unicast or multicast address and port. Defaults to: {{224,0,0,224}, 9999}

  :interface - If the host has many network interfaces, specify which one to use by IP address. Defaults to: {0,0,0,0}

  :level - the level to be logged by this backend. _Note that messages are first filtered by the general `:level` configuration in `:logger`_. Defaults to: :debug

  :format - the format message used to print logs. Defaults to: "$time $metadata[$level] $levelpad$message\n"

  :metadata - the metadata to be printed by $metadata. Defaults to: [] (no metadata)
  """

  use GenEvent
  alias Experimental.GenStage
  require GenStage
  require Logger

  @metadata Application.get_env(:logger_multicast_backend, :metadata, [])
  @format Application.get_env(:logger_multicast_backend, :format, "$time $metadata[$level] $message\n")
  @level Application.get_env(:logger_multicast_backend, :level, :debug)
  @target Application.get_env(:logger_multicast_backend, :target, {{224,0,0,224}, 9999})
  @interface Application.get_env(:logger_multicast_backend, :interface, {0,0,0,0})

  @defaults %{format: @format, metadata: @metadata, level: @level, target: @target, interface: @interface}

  @doc false
  def init(__MODULE__), do: init({__MODULE__, []})

  @doc """
  initialize the state of this logger to the configuration options provided
  """
  def init({__MODULE__, opts}) do
    #Setup state
    state =
      configure(opts, @defaults)

    #Setup GenStage Flow
    {:ok, queue} = GenStage.start_link(LoggerMulticastBackend.Queue, [])
    {:ok, sender} = GenStage.start_link(LoggerMulticastBackend.Sender, [target: state.target, interface: state.interface])
    GenStage.sync_subscribe(sender, [to: queue, min_demand: 5, max_demand: 10])

    #Add Queue Consumer to state
    state =
      state
      |> Map.merge(%{queue: queue, sender: sender})
    {:ok, state}
  end

  @doc """
  GenEvent callback to handle runtime configuration of this Backend
  """
  def handle_call({:configure, options}, %{queue: queue, sender: sender} = state) do
    state =
      configure(options, state)
      |> Map.merge(%{queue: queue, sender: sender})
    :ok = GenStage.call(sender, {:configure, state.target, state.interface})
    {:ok, :ok, state}
  end

  @doc "Add the formatted log entry to the log output queue if it meets our logging criteria"
  def handle_event({level, _gl, {Logger, message, timestamp, metadata}}, %{level: min_level} = state) do
    if is_nil(min_level) or Logger.compare_levels(level, min_level) != :lt do
      entry = format_event(level, message, timestamp, metadata, state)
      :ok = GenStage.cast(state.queue, {:enqueue, entry})
    end
    {:ok, state}
  end

  @doc """
  GenEvent callback to handle flushing of the Backend
  """
  def handle_event(:flush, state) do
    GenStage.cast(state.queue, :flush)
    {:ok, state}
  end

  # Format the event using the specified Logger, metadata and output string
  defp format_event(level, msg, ts, md, %{format: format, metadata: metadata}) do
    Logger.Formatter.format(format, level, msg, ts, take_metadata(md, metadata))
  end

  # Take specified metadata keys from provided metadata
  defp take_metadata(metadata, keys) do
    metadatas = Enum.reduce(keys, [], fn key, acc ->
      case Keyword.fetch(metadata, key) do
        {:ok, val} -> [{key, val} | acc]
        :error     -> acc
      end
    end)
    Enum.reverse(metadatas)
  end

  # Extracts configuration options from Dict and returns Map
  # Used defaults provided for any options not provided
  defp configure(opts, defaults) do
    level = Dict.get(opts, :level, defaults.level)
    target = Dict.get(opts, :target, defaults.target)
    interface = Dict.get(opts, :interface, defaults.interface)
    format = case Dict.get(opts, :format, defaults.format) do
      f when is_list(f) -> f
      f -> Logger.Formatter.compile f
    end
    metadata = Dict.get(opts, :metadata, defaults.metadata)
    %{format: format, metadata: metadata, level: level, target: target, interface: interface}
  end
end
